File: cpp/src/greedy_search.cpp
#include "greedy_search.h"

#include <algorithm>
#include <cmath>

namespace asr_decoder::internal {
namespace {

void consume_frame(DecoderState& state, int token, double probability) {
  ++state.processed_frames;
  if (token == state.config.blank_id) {
    state.last_greedy_token = -1;
    return;
  }
  if (token == state.last_greedy_token) {
    state.greedy_spans.back().second = state.processed_frames;
    state.greedy_token_log_probabilities.back() =
        std::max(state.greedy_token_log_probabilities.back(), probability);
    return;
  }
  state.greedy_tokens.push_back(token);
  state.greedy_spans.push_back(
      {state.processed_frames - 1, state.processed_frames});
  state.greedy_token_log_probabilities.push_back(probability);
  state.last_greedy_token = token;
}

DecodeResult build_result(const DecoderState& state, GreedyOptions options) {
  DecodeResult result;
  result.tokens.push_back(state.greedy_tokens);
  std::vector<Timestamp> timestamps;
  timestamps.reserve(state.greedy_spans.size());
  for (const auto& [start, end] : state.greedy_spans) {
    Timestamp timestamp{start, end};
    if (state.config.frame_shift_ms > 0.0) {
      timestamp.start_ms = start * state.config.frame_shift_ms;
      timestamp.end_ms = end * state.config.frame_shift_ms;
    }
    timestamps.push_back(timestamp);
  }
  result.timestamps.push_back(std::move(timestamps));
  if (options.return_token_probabilities) {
    std::vector<double> probabilities;
    probabilities.reserve(state.greedy_token_log_probabilities.size());
    for (double probability : state.greedy_token_log_probabilities) {
      probabilities.push_back(std::exp(probability));
    }
    result.probabilities.push_back(std::move(probabilities));
  }
  return result;
}

}  // namespace

DecodeResult run_greedy_search(DecoderState& state, const float* values,
                               size_t frames, size_t vocabulary_size,
                               GreedyOptions options) {
  DecodeResult result;
  if (!state.config_error.empty()) return result;
  if (!validate_input(state, values, frames, vocabulary_size)) return result;
  if (!lock_decode_mode(state, DecodeMode::kGreedy)) return result;

  for (size_t frame = 0; frame < frames; ++frame) {
    const float* row = values + frame * vocabulary_size;
    size_t token = 0;
    for (size_t candidate = 1; candidate < vocabulary_size; ++candidate) {
      if (row[candidate] > row[token]) token = candidate;
    }
    consume_frame(state, static_cast<int>(token), row[token]);
  }
  result = build_result(state, options);
  if (options.finalize) state.reset();
  return result;
}

}  // namespace asr_decoder::internal

File: rust/src/greedy_search.rs
use crate::decoder::DecodeMode;
use crate::{CTCDecoder, DecodeResult, DecoderError, LogProbabilities};

impl CTCDecoder {
    pub fn greedy_search(
        &mut self,
        input: LogProbabilities<'_>,
        finalize: bool,
        return_token_probabilities: bool,
    ) -> Result<DecodeResult, DecoderError> {
        self.validate_input(input)?;
        self.lock_decode_mode(DecodeMode::Greedy)?;
        for frame_index in 0..input.frames() {
            let row = input.frame(frame_index);
            let mut token_id = 0;
            for candidate in 1..row.len() {
                if row[candidate] > row[token_id] {
                    token_id = candidate;
                }
            }
            let score = f64::from(row[token_id]);
            self.consume_greedy_frame(token_id, score);
        }

        let result = DecodeResult {
            tokens: vec![self.greedy_tokens.clone()],
            timestamps: vec![self.timestamps(&self.greedy_spans)],
            probabilities: return_token_probabilities.then(|| {
                vec![self
                    .greedy_token_log_probabilities
                    .iter()
                    .map(|score| score.exp())
                    .collect()]
            }),
            scores: None,
            gating: None,
        };
        if finalize {
            self.reset();
        }
        Ok(result)
    }

    fn consume_greedy_frame(&mut self, token_id: usize, score: f64) {
        self.processed_frames += 1;
        if token_id == self.blank_id {
            self.last_greedy_token = None;
            return;
        }
        if self.last_greedy_token == Some(token_id) {
            let span = self
                .greedy_spans
                .last_mut()
                .expect("a repeated token has an existing span");
            span.1 = self.processed_frames;
            let probability = self
                .greedy_token_log_probabilities
                .last_mut()
                .expect("a repeated token has an existing probability");
            *probability = probability.max(score);
            return;
        }
        self.greedy_tokens.push(token_id);
        self.greedy_spans
            .push((self.processed_frames - 1, self.processed_frames));
        self.greedy_token_log_probabilities.push(score);
        self.last_greedy_token = Some(token_id);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::DecoderConfig;

    #[test]
    fn collapses_repeats_and_blank_separated_tokens() {
        let values = [
            -5.0, 0.0, -5.0, -5.0, 0.0, -5.0, 0.0, -5.0, -5.0, -5.0, 0.0, -5.0,
        ];
        let input = LogProbabilities::new(&values, 4, 3).unwrap();
        let mut decoder = CTCDecoder::new(DecoderConfig {
            frame_shift_ms: Some(20.0),
            ..DecoderConfig::default()
        })
        .unwrap();
        let result = decoder.greedy_search(input, true, true).unwrap();
        assert_eq!(result.tokens, vec![vec![1, 1]]);
        assert_eq!(result.timestamps[0][0].end_frame, 2);
        assert_eq!(result.timestamps[0][1].start_ms, Some(60.0));
    }
}
